File: week 1/hwk1.R
#data:
adults    <- c(2.4,3.2,3.9,5.7,6,7.4,8.2,10,10.1,10.4,11.3,12.8,18,24)
juveniles <- c(11.6,7.1,14.3,19.1,12.4,19.7,31.5,18.5,22.1,26.9,19.2,21,18.1,26.8)
temps     <- c(12.00,15.04,13.45,12.03,13.00,13.48,9.24,15.33,12.08,9.50,17.10,12.16,22.00,13.64)
data      <- data.frame(adults=adults, juveniles=juveniles, temps=temps)
Nobs      <- nrow(data)
plot(data)

File: week 1/week 1 soln/hwk1.cpp
#include <TMB.hpp>
/* Nonlinear regression example
*  Juveniles = (alpha*Adults) / (1 + beta*Adults) + c*temp + eps
*  where eps ~ N(0, SD)
*  
*  Cahill Oct 2019
*/
template<class Type>
Type objective_function<Type>::operator() ()
{
  // Data
  DATA_VECTOR( Y_i );  //measurement of juveniles i
  DATA_VECTOR( A_i );  //measurement of adults i
  DATA_VECTOR( T_i );  //measurement of temp i   
  
  // Parameters
  PARAMETER_VECTOR( parm_vec ); 
  
  Type alpha = parm_vec(0); 
  Type beta = parm_vec(1); 
  Type c_temp = parm_vec(2); 
  Type sigma = exp(parm_vec(3)); //sigma cannot be negative

  int Nobs = Y_i.size();
  vector<Type> Y_pred_i( Nobs ); 
  
  // Objective function
  Type nll = 0;
  
  for( int i=0; i<Nobs; i++ ){
    Y_pred_i(i) = (alpha*A_i(i)) /( (Type(1.0) + beta*A_i(i)) ) + c_temp*T_i(i); 
    nll -= dnorm( Y_i(i), Y_pred_i(i), sigma, true ); 
  }
  
  SIMULATE{
   Y_i = rnorm(Y_pred_i, sigma); 
   REPORT(Y_i); 
  }
  return nll;
}

